src/main/java/org/broadinstitute/hellbender/tools/exome/ReadGroup.java

```diff
@@ -0,0 +1,164 @@
+*  7.4 Notice. Any notices under this Agreement shall be in writing, shall specifically refer to this Agreement, and shall be sent by hand, recognized national overnight courier, confirmed facsimile transmission, confirmed electronic mail, or registered or certified mail, postage prepaid, return receipt requested.  All notices under this Agreement shall be deemed effective upon receipt.
+*  7.5 Amendment and Waiver; Entire Agreement. This Agreement may be amended, supplemented, or otherwise modified only by means of a written instrument signed by all parties. Any waiver of any rights or failure to act in a specific instance shall relate only to such instance and shall not be construed as an agreement to waive any rights or fail to act in any other instance, whether or not similar. This Agreement constitutes the entire agreement among the parties with respect to its subject matter and supersedes prior agreements or understandings between the parties relating to its subject matter.
+*  7.6 Binding Effect; Headings. This Agreement shall be binding upon and inure to the benefit of the parties and their respective permitted successors and assigns. All headings are for convenience only and shall not affect the meaning of any provision of this Agreement.
+*  7.7 Governing Law. This Agreement shall be construed, governed, interpreted and applied in accordance with the internal laws of the Commonwealth of Massachusetts, U.S.A., without regard to conflict of laws principles.
+*/
+package org.broadinstitute.hellbender.tools.exome;
+
+/**
+ * Class that represent a read-group.
+ *
+ * @author Valentin Ruano-Rubio &lt;valentin@broadinstitute.org&gt;
+ */
+class ReadGroup {
+
+    private final String id;
+
+    /**
+     * Sets the read-group sample.
+     *
+     * <p>
+     *     This method will fail with an exception if the read-group has already a sample assigned different
+     *     to the sample provided as an input or if the latter is {@code null}.
+     * </p>
+     *
+     * <p>
+     *     This is done to prevent false sharing of a read-group across samples.
+     * </p>
+     *
+     * @param sample the new sample for the read-group.
+     *
+     * @throws IllegalStateException if this read-group has already a sample and the sample provide is another.
+     *
+     */
+    protected void setSample(final Sample sample) {
+        if (this.sample != null && !this.sample.equals(sample))
+            throw new IllegalStateException("attempt to set a sample-full read-group with another sample");
+        this.sample = sample;
+    }
+
+    /**
+     * Unconditional and unchecked change of the sample for this read-group.
+     *
+     * @param newSample the new sample (can be null).
+     */
+    protected void resetSample(final Sample newSample) {
+        this.sample = newSample;
+    }
+
+    private Sample sample;
```

Comment:
aha! a mutable field. this make this class even more suspect. Mutable classes should not be treated as values (and the presence of hashcode and equals indicates a value.